test/box-luatest/gh_10032_update_upsert_splice_test.lua

```diff
@@ -0,0 +1,274 @@
+                res.ret = res.ret:totable()
+            else
+                res.ret = table.copy(res.ret)
+            end
+            if varbinary.is(res.ret[2]) then
+                res.ret[2] = '!!bin ' .. tostring(res.ret[2])
+            end
+        end
+        return res
+    end
+    return {
+        ops = test_case.ops,
+        expect_res = prettify_case_res(test_case.res),
+        got_res = prettify_case_res(got_res),
+    }
+end
+]]
+local get_test_case_message = loadstring(get_test_case_message_source)()
+
+-- Get field type - 'bin' or 'str'.
+-- This function is required both on host and on server, and unfortunately now
+-- test engine cannot pass functions as exec arguments, so define source code.
+local field_type_source = [[
+return function(field)
+    local varbinary = require('varbinary')
+    if type(field) == 'string' then
+        return 'str'
+    elseif varbinary.is(field) then
+        return 'bin'
+    end
+end
+]]
+local field_type = loadstring(field_type_source)()
+
+-- Generate array with 2nd field of given type ('str' or 'bin').
+local function generate_tuple_template(data_type)
+    assert(data_type == 'str' or data_type == 'bin')
+    local tmpl = {100, '1234', 200}
+    if data_type == 'bin' then
+        tmpl[2] = varbinary.new(tmpl[2])
+    end
+    return tmpl
+end
+
+-- Test lots of various variants of splice (':') of tuple update.
+g.test_tuple_update_splice = function(cg)
+    local tmpl = generate_tuple_template(cg.params.type)
+    local tuple = box.tuple.new(tmpl)
+    local test_cases = generate_test_cases(tmpl)
+
+    for _, test_case in ipairs(test_cases) do
+        local ok, ret = pcall(tuple.update, tuple, test_case.ops)
+        ret = ok and ret or tostring(ret)
+        local res = {ok = ok, ret = ret}
+        local msg = get_test_case_message(test_case, res)
+        t.assert_equals(res, test_case.res, msg)
+        if ok then
+            t.assert_equals(field_type(ret[2]), cg.params.type, msg)
+        end
+    end
+end
+
+-- Test lots of various variants of splice (':') of memtx space update.
+g.test_memtx_update_splice = function(cg)
+    local tmpl = generate_tuple_template(cg.params.type)
+    local test_cases = generate_test_cases(tmpl)
+
+    cg.server:exec(function(data_type, tmpl, test_cases,
+                            get_test_case_message_source, field_type_source)
+        local get_test_case_message = loadstring(get_test_case_message_source)()
+        local field_type = loadstring(field_type_source)()
+
+        local s = box.schema.space.create('test', {engine = 'memtx'})
+        s:create_index('pk')
+
+        for _, test_case in ipairs(test_cases) do
+            local ops = test_case.ops
+            s:replace(tmpl)
+            local ok, ret = pcall(s.update, s, tmpl[1], ops)
+            ret = ok and ret or tostring(ret)
+            local res = {ok = ok, ret = ret}
+            local msg = get_test_case_message(test_case, res)
```

Comment:
I guess it is easier to consider failure and success cases explicitly:

```lua
    local ok, ret = pcall(<>)
    local ops = test_case.ops
    t.assert_equals(test_case.ok, ok, ops)
    if ok then
        t.assert_equals(ret, test_case.ret, ops)
        t.assert_equals(field_type, data_type, ops)
    else
        t.assert_equals(tostring(ret), test_case.ret, ops)
    end
```

Reply:
I don't see much difference, but ok.